includes/validation/class-amp-validation-error-taxonomy.php

```diff
@@ -2086,6 +2185,163 @@ public static function render_single_url_error_details( $validation_error, $term
+		$plugins = get_plugins();
+		if ( isset( $plugins[ $plugin_slug ] ) ) {
+			return [
+				'name' => $plugin_slug,
+				'data' => $plugins[ $plugin_slug ],
+			];
+		}
+		foreach ( $plugins as $plugin_file => $plugin_data ) {
+			if ( strtok( $plugin_file, '/' ) === $plugin_slug ) {
+				return [
+					'name' => $plugin_file,
+					'data' => $plugin_data,
+				];
+			}
+		}
+		return null;
+	}
+
+	/**
+	 * Render sources.
+	 *
+	 * @param array $sources Sources.
+	 */
+	private static function render_sources( $sources ) {
+		?>
+		<details>
+			<summary>
+				<?php
+				$source_count = count( $sources );
+				echo esc_html(
+					sprintf(
+						/* translators: %s: number of sources. */
+						_n(
+							'%s source',
+							'%s sources',
+							$source_count,
+							'amp'
+						),
+						number_format_i18n( $source_count )
+					)
+				);
+				?>
+			</summary>
+			<ol class="validation-error-sources">
+				<?php foreach ( $sources as $source ) : ?>
+					<?php
+					$file_text = null;
+					$edit_url  = null;
+					if ( isset( $source['file'], $source['line'] ) ) {
+						$file      = $source['file'];
+						$line      = $source['line'];
+						$file_text = "$file:$line";
+
+						if ( isset( $source['type'], $source['name'] ) ) {
+							if ( 'plugin' === $source['type'] && current_user_can( 'edit_plugins' ) ) {
+								$plugin = self::get_plugin_from_slug( $source['name'] );
+								if ( $plugin ) {
+									// Prepend the plugin directory name to the file name as the plugin editor requires.
+									if ( false !== strpos( $plugin['name'], '/' ) ) {
+										$file = strtok( $plugin['name'], '/' ) . '/' . $file;
+									}
+
+									$edit_url = add_query_arg(
+										[
+											'plugin' => rawurlencode( $plugin['name'] ),
+											'file'   => rawurlencode( $file ),
+											'line'   => rawurlencode( $line ),
+										],
+										admin_url( 'plugin-editor.php' )
+									);
+								}
+							} elseif ( 'theme' === $source['type'] && current_user_can( 'edit_themes' ) ) {
+								$edit_url = add_query_arg(
+									[
+										'file'  => rawurlencode( $file ),
+										'theme' => rawurlencode( $source['name'] ),
+										'line'  => rawurlencode( $line ),
+									],
+									admin_url( 'theme-editor.php' )
+								);
+							}
+						}
+
+						/**
+						 * Filters the URL for opening the file to edit.
+						 *
+						 * Users of IDEs that support opening files in via web protocols can use this filter to override
+						 * the edit link to result in their editor opening rather than the theme/plugin editor.
+						 *
+						 * @since 1.4
+						 *
+						 * @param string|null $edit_url Edit URL for the file. Line number with colon will be appended to end. Null when user cannot access theme/plugin editor.
+						 * @param array       $source   Source information.
+						 */
+						$edit_url = apply_filters( 'amp_validation_error_source_file_edit_url', $edit_url, $source );
```

Comment:
@schlessera Is this what you had in mind?

Reply:
Yes, awesome. This way, developers can already hook into this and just point it to whatever their editor happens to be.

We can add automagic later on to be smart about this, but that's more about discovery then. The filter is already enough to allow this to work.

As an example, if I now filter this to produce `phpstorm://open?file=%file&line=%line`, it should automatically work on my Mac to open the files in PHPStorm.

Reply:
How's this? e6cbc522

Updated example plugin for opening files in an external editor: https://gist.github.com/westonruter/1a36a78a79fb1abbdfd1bb5b38a9f3a1